src/Deserializer.cpp

```diff
@@ -16,15 +16,327 @@
+
+std::string Deserializer::version() {
+  return _dgVersion;
+}
+
+std::string Deserializer::preview() {
+  return _preview;
+}
+
+std::string Deserializer::roomName() {
+  return _roomName;
+}
+
+////////////////////////////////////////////////////////////
+// Implementation - State changes
+////////////////////////////////////////////////////////////
+
+bool Deserializer::readHeader() {
+  { // Read Dagon version string
+    uint8_t len;
+    if (!(len = SDL_ReadU8(_rw)))
+      return false;
+    std::vector<uint8_t> buf(len);
+    if (!SDL_RWread(_rw, buf.data(), buf.capacity(), 1))
+      return false;
+    _dgVersion = std::string(buf.begin(), buf.end());
+  }
+
+  { // Read preview text
+    uint8_t len;
+    if (!(len = SDL_ReadU8(_rw)))
+      return false;
+    std::vector<uint8_t> buf(len);
+    if (!SDL_RWread(_rw, buf.data(), buf.capacity(), 1))
+      return false;
+    _preview = std::string(buf.begin(), buf.end());
+  }
+
+  { // Read Room name
+    uint8_t len;
+    if (!(len = SDL_ReadU8(_rw)))
+      return false;
+    std::vector<uint8_t> buf(len);
+    if (!SDL_RWread(_rw, buf.data(), buf.capacity(), 1))
+      return false;
+    _roomName = std::string(buf.begin(), buf.end());
+  }
+
+  return true;
+}
+
+bool Deserializer::readScriptData() {
+  uint32_t numFields = SDL_ReadBE32(_rw);
+  for (uint32_t i = 0; i < numFields; i++) {
+    uint16_t lineLen = SDL_ReadBE16(_rw);
+    std::vector<uint8_t> lineBuf(lineLen);
+    if (!SDL_RWread(_rw, lineBuf.data(), lineBuf.capacity(), 1))
+      return false;
+
+    std::string line(lineBuf.begin(), lineBuf.end());
+    int errCode = luaL_dostring(_L, line.c_str());
+    if (errCode != 0) { // Perhaps one of the tables we are inserting into doesn't exist
+      // TODO: begin creating empty tables at the deepest level of the line
+    }
+  }
+
+  return true;
+}
+
+void Deserializer::toggleSpots() {
+  Room *room = Control::instance().currentRoom();
+
+  uint16_t numNodes = SDL_ReadBE16(_rw);
+  if (numNodes != room->numNodes()) {
+    Log::instance().warning(kModScript, "Node number mismatch. Expected %d. Got %d", numNodes,
+                            room->numNodes());
+  }
+
+  // Restore spot activation statuses
+  if (room->hasNodes()) {
+    room->beginIteratingNodes();
+
+    do {
+      Node *node = room->iterator();
+
+      uint16_t numSpots = SDL_ReadBE16(_rw);
+      if (numSpots != node->numSpots()) {
+        Log::instance().warning(kModScript, "Spot number mismatch. Expected %d. Got %d", numSpots,
+                                node->numSpots());
+      }
+
+      if (node->hasSpots()) {
+        node->beginIteratingSpots();
+
+        do {
+          Spot *spot = node->currentSpot();
+          bool enable = SDL_ReadU8(_rw);
+          if (enable)
+            spot->enable(true); // TODO: find out if we should always force enable.
+          else
+            spot->disable(true); // TODO: find out if we should always force disable.
```

Comment:
Well, @agustincordes  should we always force the enable/disable? My rationale for this decision so far has been that it seems that by forcing, I affect the spot individually and since spot statuses are saved individually I guessed this would be correct.